### ISSUE TITLE
MRC and MCR instructions are not supported

### ISSUE THREAD

**Issue by valour01** [angr/angr#854]
Hi. 
When I used the `proj.factory.block(addr)` to create blocks. I noticed some behavior below:

For an elf file whose architecture is ARM. There might be some command angr might not understand. For example `MRC, MCR`. I noticed that angr will indicate an end of a block when it come across such kinds of instructions. 

Is it a right behavior? Or where I can patch this so that angr will not treat these instructions as the split of different blocks.

**Comment by ltfish** [comment]
> Is it a right behavior?

That is the right behavior - I believe MRC and MCR are co-processor instructions, and VEX simply does not support them. angr will definitely stop decoding the block when there is anything undecodable.

> Or where I can patch this so that angr will not treat these instructions as the split of different blocks.

I believe now there is a way (a new and _advanced_ feature in PyVEX or gymrat) to write your own lifter, that explicitly lifts/patches byte sequences that VEX does not support otherwise. I think there is some documentation available, but I'm not sure where and how complete it is, and I really don't have time to look it up and post the link here right now. I would recommend you to take a look at [PyVEX](https://github.com/angr/pyvex) and [angr-platforms](https://github.com/angr/angr-platforms), where gymrat is implemented, and where we wrote lifters for new architectures in Python. Modeling the behavior of `MRC` or `MCR` is a different story, but at least you can nop them out ;)

The above is the correct way to solve this problem, but it might be take too much time to get started if you are new to angr, PyVEX, and angr-platforms. An alternative approach (which is a giant hack) is to manually patch the byte sequences of `MRC` and `MCR` in the binary to `nop` before lifting. You can do this using a hex editor and modify the binary, or you can modify the content in the memory backend (`proj.loader.main_object.mem` or something like that, I don't quite remember the full thing).

@Nilocunger might have something to say on this issue, too!

**Comment by rhelmot** [comment]
Note - you can detect blocks that are split like this by examining `block.jumpkind` - it will be `"Ijk_NoDecode"` when there's been a decoding error.

There's no easy way to get the block to continue across unknown instructions besides adding an extension as fish described. You could probably just get away with modeling them as NOPs for a first pass.

**Comment by valour01** [comment]
@ltfish  Thanks for your quick reply. If I understand right, I need to add the extension on the VEX IR to support the ARM co-processor instructions. Am I right? I will try to work on these. Besides, does valgrind  support the ARM co-processor instructions? If so, why we cannot move those support to angr. If not, I will work on my own.

@rhelmot Thanks for your  comment. However, I find `block.jumpkind` does not work. I use the code below

```
block = p.factory.block(address)
block.jumpkind
```

However, It doesn't work. Could you please to give a simple example code? Many Thanks
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-7-3d4d8d41c600> in <module>()
----> 1 block.jumpkind

AttributeError: 'Block' object has no attribute 'jumpkind'
```

**Comment by rhelmot** [comment]
Whoops. Not sure why that's not available. You should use `block.vex.jumpkind`.

**Comment by Nilocunger** [comment]
To use Gymrat to nop out that instruction, all you'll have to do is write a new lifter and register it. There are some good examples in [the pyvex spotters test](https://github.com/angr/pyvex/blob/master/tests/test_spotter.py). For reference, here is the easiest to understand example:
```python
class Instruction_IMAGINARY(Instruction):
    bin_format = bin(0x45e1)[2:].zfill(16)
    name = 'IMAGINARY'

    def compute_result(self):
        a = self.constant(10, Type.int_27)
        b = self.constant(20, Type.int_27)
        a + b

class ImaginarySpotter(GymratLifter):
    instrs = [Instruction_IMAGINARY]

register(ImaginarySpotter, 'ARMEL')
```
You'll create a class for each instruction you want to implement and give them each a binary format `bin_format` and name. Then instead of having the `compute_result` function do something, you can nop out the instruction by replacing the body with `pass`. Then all you have to do is create a lifter class with the instrs and register it for your architecture. This way, when libVEX hits an instruction it can't decode, your lifter will be run, match the instruction, and lift it. Documentation for `bin_format` can be found [here](https://github.com/angr/pyvex/blob/master/pyvex/lift/util/instr_helper.py). Let me know you have any questions!

**Comment by ltfish** [comment]
> Besides, does valgrind support the ARM co-processor instructions? If so, why we cannot move those support to angr. If not, I will work on my own.

We would have moved the support to angr (or, directly using the support from angr) if it is supported in libVEX. So the answer is no (unless I'm wrong and it's supported in new versions of Valgrind). You are welcome to check the latest version of Valgrind, but I think it's very likely that you'll have to add the support by yourself.

As far as I know, we also do not model co-processors at all. You'll need to add quite a bit code if you want to do anything more than just nopping out those instructions.

**Comment by valour01** [comment]
@Nilocunger Thank you so much for your examples. Though I am not very familiar with the `bin_format`. I know how those functions work now. I don't want to simply nop out the instruction. I will try to model the co-processor instructions and hopefully I could do some contributions. 

I noticed that there are many abstract functions inside the class Instruction. Maybe `bin_format` and `compute_result` are two of the most important features. However, I need to know when I try to model one new instruction. Where should I start(Here I mean, where should I put my sub-class in), which functions do I need to re-write except `compute_result`. I tried to find an example. However, I failed. I want to know how pyvex modeled some normal instructions like `load, store, bl, etc`in ARM. I guess every specific instruction will map to one sub-class. Where is the source code? Many Thanks

@ltfish Thanks for your comments. I will try to implement it and let angr to support co-processor related instructions :).

**Comment by valour01** [comment]
@Nilocunger After doing some debug work. I noticed that the lift process is done by the function `vex_lift` defined in the file `pyvex/pyvex.c`. I am not sure whether I am right. If so, is it mean I need to rewrite the native code? I don't know where should I write if I write python code to extend the instruction by creating class inheriting the`Instruction`. 

It seems  that we can re-write libVEX supported by valgrind or create sub-class inheriting the `Instruction` to do the extension. What are the differences. Maybe I am not very familiar with lift process angr adopt. It would be great if you can give me some hint. Many Thanks

**Comment by ltfish** [comment]
> I am not sure whether I am right. If so, is it mean I need to rewrite the native code?

The new features in PyVEX (gymrat) allow you to write a lifter in pure Python that complements the existing lifter. See the previous comment that @Nilocunger provided with Python code snippets. You are really not supposed to touch `pyvex.c` unless you *really* want to.

**Comment by subwire** [comment]
This is very much solved in #1234 and friends. 
However, since we have no support for Coprocs, we probably should, for the sake of accuracy, do something a little different than we do now. See pyvex/pyvex/lifting/gym/arm_spotter.py for exactly what should be added by anyone that needs it.
